include/circt/Dialect/HW/HWTypes.td

```diff
@@ -137,3 +137,30 @@ def UnionType : HWType<"Union"> {
     mlir::Type getFieldType(mlir::StringRef fieldName);
   }];
 }
+
+def TypeAliasType : HWType<"TypeAlias"> {
+  let summary = "An symbolic reference to a type declaration";
+  let description = [{
+    A TypeAlias is parameterized by a SymbolRefAttr, which points to a
+    TypedeclOp. The root referece should refer to a TypeScope within the same
+    outer ModuleOp, and the leaf reference should refer to a type within that
+    TypeScope. A TypeAlias is further parameterized by the inner type, which is
+    needed to be known at the time the type is parsed.
+
+    Upon construction, a TypeAlias stores the symbol reference and type, and
+    canonicalizes the type to resolve any nested type aliases. The canonical
+    type is also stored to avoid recomputing it when needed.
+  }];
+
+  let mnemonic = "typealias";
+
+  let parameters = (ins
+    "mlir::SymbolRefAttr":$ref,
+    "mlir::Type":$innerType,
+    "mlir::Type":$canonicalType
+  );
+
+  let extraClassDeclaration = [{
+    static TypeAliasType get(SymbolRefAttr ref, Type type);
```

Comment:
I'm pretty sure this should be modeled as a ["custom type builder method"](https://mlir.llvm.org/docs/OpDefinitions/#custom-type-builder-methods). This is a relatively new feature in MLIR.